btor/src/boolector_solver.cpp

```diff
@@ -200,11 +210,31 @@ void BoolectorSolver::assert_formula(const Term & t)
 {
   std::shared_ptr<BoolectorTerm> bt =
       std::static_pointer_cast<BoolectorTerm>(t);
-  boolector_assert(btor, bt->node);
+  if (produce_unsat_cores)
+  {
+    // unsat core production needs to use assumptions
+    unsat_core_assumptions[unsat_core_assumptions.size() - 1].push_back(
```

Comment:
i don't understand this

Reply:
That isn't clear, you're right. Added a comment with: fa8376eef19d36053e7105c6a0fa93d5bed5e721

Reply:
OK. what is the assumption here?

Reply:
Right, so there's an implicit assumption that `unsat_core_assumptions` has one more elements than the current context level. And then, we just need to use `unsat_core_assumptions` because Boolector gets unsat cores using assumptions and those get removed after each call to `check_sat` so we need to maintain a list of the assertions to add before checking sat. This is one of the reasons I wanted you to review this actually. Do you think there's a better way to implement this when it depends on assumptions? With this implementation, we could lose performance because of solving with assumptions instead of assertions.